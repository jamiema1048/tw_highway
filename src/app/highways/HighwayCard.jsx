import { useState, useRef } from "react";
import Link from "next/link";

const HighwayCard = ({
  highway,
  hoveredHighway,
  setHoveredHighway,
  loading,
  setLoading,
  error,
  setError,
}) => {
  const timeoutRef = useRef(null);
  let mutex = Promise.resolve();

  // 處理手機長按
  const handleTouchStart = (highway) => {
    if (touchTimeout) clearTimeout(touchTimeout);
    const timeout = setTimeout(() => {
      setHoveredHighway(highway);
    }, 500); // 0.5秒後才觸發，避免短按立即觸發
    setTouchTimeout(timeout);
  };

  // 取消長按事件（短按無效）
  const handleTouchEnd = () => {
    if (touchTimeout) clearTimeout(touchTimeout);
  };

  // 這部分可以忽略，專注於字卡和圖片切換邏輯
  const handleNextImage = () => {
    if (
      hoveredHighway &&
      hoveredHighway.images &&
      hoveredHighway.currentImageIndex < hoveredHighway.images.length - 1
    ) {
      setHoveredHighway((prev) => ({
        ...prev,
        currentImageIndex: prev.currentImageIndex + 1,
      }));
    }
  };

  const handlePrevImage = () => {
    if (hoveredHighway && hoveredHighway.currentImageIndex > 0) {
      setHoveredHighway((prev) => ({
        ...prev,
        currentImageIndex: prev.currentImageIndex - 1,
      }));
    }
  };
  const handleMouseEnter = () => {
    if (timeoutRef.current) {
      clearTimeout(timeoutRef.current);
    }
    // async function writeCard() {
    //   // 只有當 hoveredHighway 不同時才更新狀態，避免不必要的重渲染
    //   if (!hoveredHighway || hoveredHighway.id !== highway.id) {
    //     await Promise.all([setHoveredHighway(highway)]);
    //   }
    // }
    // writeCard();
    timeoutRef.current = setTimeout(() => {
      // 只有當 hoveredHighway 不同時才更新狀態，避免不必要的重渲染
      if (!hoveredHighway || hoveredHighway.id !== highway.id) {
        setHoveredHighway(highway);
      }
    }, 300);
  };

  const handleMouseLeave = (event) => {
    timeoutRef.current = setTimeout(() => {
      // 如果滑鼠進入其他 Link，則不關閉字卡
      const relatedTarget = event.relatedTarget;
      if (
        relatedTarget &&
        (relatedTarget.closest("a") || relatedTarget.closest(".highway-card"))
      ) {
        return;
      }

      async function clearCard() {
        await Promise.all([setHoveredHighway(null)]);
      }
      clearCard();
    }, 300);
  };

  const handleCardMouseEnter = () => {
    if (timeoutRef.current) {
      clearTimeout(timeoutRef.current);
    }
  };

  const handleCardMouseLeave = () => {
    timeoutRef.current = setTimeout(() => {
      setHoveredHighway(null);
    }, 300);
  };

  return (
    <h3 className="relative text-xl">
      <Link
        href={`highways/${highway.id}`}
        className="font-bold text-white-600 hover:text-yellow-400 active:text-yellow-600"
        onMouseEnter={handleMouseEnter}
        onMouseLeave={handleMouseLeave}
        onTouchStart={handleTouchStart}
        onTouchEnd={handleTouchEnd}
      >
        {highway.remark && highway.remark.includes("解編")
          ? `${highway.name}  (已解編)`
          : highway.remark && highway.remark.includes("未納編")
          ? `${highway.name}  (未納編)`
          : highway.name}
      </Link>

      {hoveredHighway?.id === highway.id && (
        <div
          className="absolute bottom-full left-36 transform -translate-x-1/2 mb-4 w-80 p-4 bg-white border border-gray-300 shadow-lg rounded-lg z-50 highway-card"
          onMouseEnter={handleCardMouseEnter}
          onMouseLeave={handleCardMouseLeave}
        >
          {/* 如果有多張圖片 */}
          {hoveredHighway.images && hoveredHighway.images.length > 1 ? (
            <div className="relative">
              {/* 顯示圖片 */}
              <img
                src={hoveredHighway.images[hoveredHighway.currentImageIndex]}
                alt={hoveredHighway.name}
                className="w-full h-48 object-cover rounded-md"
                style={{ objectPosition: "center" }} // 設置圖片顯示中間部分
              />

              {/* 只有當圖片索引大於0時，顯示上一張按鈕 */}
              <button
                className={`absolute top-1/2 left-0 transform ${
                  hoveredHighway.currentImageIndex > 0 ? "block" : "hidden"
                } -translate-y-1/2 bg-gray-800 text-white p-2 rounded-full z-10`}
                onClick={handlePrevImage}
                disabled={hoveredHighway.currentImageIndex <= 0}
              >
                ◀
              </button>

              {/* 只有當圖片索引小於圖片數量減一時，顯示下一張按鈕 */}
              <button
                className={`absolute top-1/2 right-0 transform ${
                  hoveredHighway.currentImageIndex <
                  hoveredHighway.images.length - 1
                    ? "block"
                    : "hidden"
                } -translate-y-1/2 bg-gray-800 text-white p-2 rounded-full z-10`}
                onClick={handleNextImage}
                disabled={
                  hoveredHighway.currentImageIndex >=
                  hoveredHighway.images.length - 1
                }
              >
                ▶
              </button>
            </div>
          ) : hoveredHighway.images && hoveredHighway.images.length === 1 ? (
            // 單張圖片顯示
            <div className="w-full h-48 flex items-center justify-center bg-gray-200 rounded-md">
              <img
                src={hoveredHighway.images[0]}
                alt={hoveredHighway.name}
                className="w-full h-48 object-cover rounded-md"
              />
            </div>
          ) : (
            // 無圖片顯示
            <div className="w-full h-48 flex items-center justify-center bg-gray-200 rounded-md">
              <span className="text-gray-500">No image to show</span>
            </div>
          )}
          {hoveredHighway.images && hoveredHighway.images.length > 0 ? (
            <p className="text-sm text-black font-semibold mt-2">
              {hoveredHighway.currentImageIndex + 1}/
              {hoveredHighway.images.length}
            </p>
          ) : (
            <></>
          )}
          <p className="text-sm text-black font-semibold mt-2">
            {highway.name}
          </p>
          <p className="text-sm text-black font-semibold mt-1">
            {highway.routeName}
          </p>
        </div>
      )}
    </h3>
  );
};

export default HighwayCard;
