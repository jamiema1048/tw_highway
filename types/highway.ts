// types/highway.ts
export interface Highway {
  id: string;
  name: string;
  prefix?: string;
  description?: string;
  images?: string[];
  currentImageIndex?: number;
  theMostCorres?: string;
}
