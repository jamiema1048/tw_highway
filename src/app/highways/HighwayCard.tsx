"use client";

import { useState, useRef, MouseEvent, TouchEvent, useEffect } from "react";
import Link from "next/link";
import { Highway } from "types/highway";

interface Props {
  highway: Highway;
  hoveredHighway: Highway | null;
  setHoveredHighway: (hwy: Highway | null) => void;
}

interface TooltipPosition {
  top: number;
  left: number;
}

export default function HighwayCard({
  highway,
  hoveredHighway,
  setHoveredHighway,
}: Props) {
  const timeoutRef = useRef<NodeJS.Timeout | null>(null);
  const [touchTimeout, setTouchTimeout] = useState<NodeJS.Timeout | null>(null);
  const cardRef = useRef<HTMLDivElement | null>(null);
  const [position, setPosition] = useState<"top" | "bottom">("top"); // üëà Êñ∞Â¢û state
  const [visible, setVisible] = useState(false);
  const [showTooltip, setShowTooltip] = useState(false);

  // Áï∂ hover ËÆäÂåñÊôÇÔºåÊ±∫ÂÆöÂ≠óÂç°‰ΩçÁΩÆ‰∏¶ÊéßÂà∂È°ØÁ§∫
  useEffect(() => {
    if (hoveredHighway?.id === highway.id && cardRef.current) {
      const rect = cardRef.current.getBoundingClientRect();

      // Â∑¶Âè≥Ë∂ÖÁïåËôïÁêÜ
      if (rect.right > window.innerWidth) {
        cardRef.current.style.left = "auto";
        cardRef.current.style.right = "0";
        cardRef.current.style.transform = "none";
      }
      if (rect.left < 0) {
        cardRef.current.style.left = "0";
        cardRef.current.style.transform = "none";
      }

      // ‰∏ä‰∏ãË∂ÖÁïåËôïÁêÜ
      if (rect.top < 0) {
        setPosition("bottom");
      } else if (rect.bottom > window.innerHeight) {
        setPosition("top");
      } else {
        setPosition("top");
      }

      // ‰∏ã‰∏ÄÂπÄÊâçÈ°ØÁ§∫ ‚Üí ÈÅøÂÖçÂÖàÂá∫ÁèæÈåØ‰Ωç
      requestAnimationFrame(() => setVisible(true));
    } else {
      setVisible(false);
    }
  }, [hoveredHighway, highway.id]);

  // Touch ÊâãÂã¢Ëß∏Áôº hover
  const handleTouchStart = () => {
    if (touchTimeout) clearTimeout(touchTimeout);
    const timeout = setTimeout(() => {
      setHoveredHighway(highway);
      setShowTooltip(true);
    }, 500);
    setTouchTimeout(timeout);
  };

  const handleTouchEnd = () => {
    if (timeoutRef.current) {
      clearTimeout(timeoutRef.current);
      timeoutRef.current = null;
    }
    setShowTooltip(false);
    setHoveredHighway(null);
  };

  // ÂàáÊèõÂúñÁâá
  const handleNextImage = () => {
    if (
      hoveredHighway?.images &&
      hoveredHighway.currentImageIndex! < hoveredHighway.images.length - 1
    ) {
      setHoveredHighway({
        ...hoveredHighway,
        currentImageIndex: (hoveredHighway.currentImageIndex || 0) + 1,
      });
    }
  };

  const handlePrevImage = () => {
    if (hoveredHighway && (hoveredHighway.currentImageIndex || 0) > 0) {
      setHoveredHighway({
        ...hoveredHighway,
        currentImageIndex: (hoveredHighway.currentImageIndex || 0) - 1,
      });
    }
  };

  // Mouse hover
  const handleMouseEnter = () => {
    if (timeoutRef.current) clearTimeout(timeoutRef.current);
    timeoutRef.current = setTimeout(() => {
      setHoveredHighway(highway);
      setShowTooltip(true);
    }, 300);
  };

  const handleMouseLeave = (event: MouseEvent<HTMLAnchorElement>) => {
    timeoutRef.current = setTimeout(() => {
      const relatedTarget = event.relatedTarget as HTMLElement | null;
      if (
        relatedTarget &&
        (relatedTarget.closest("a") || relatedTarget.closest(".highway-card"))
      )
        return;
      setShowTooltip(false);
      setHoveredHighway(null);
    }, 300);
  };

  const handleCardMouseEnter = () => {
    if (timeoutRef.current) clearTimeout(timeoutRef.current);
  };

  const handleCardMouseLeave = () => {
    timeoutRef.current = setTimeout(() => {
      setHoveredHighway(null);
    }, 300);
  };

  return (
    <h3 className="relative text-xl">
      <Link
        href={`highways/${highway.id}`}
        className="font-bold text-white-600 hover:text-yellow-400 active:text-yellow-600 highway-link"
        onMouseEnter={handleMouseEnter}
        onMouseLeave={handleMouseLeave}
        onTouchStart={handleTouchStart}
        onTouchEnd={handleTouchEnd}
      >
        {highway.remark?.includes("Ëß£Á∑®")
          ? `${highway.name} (Â∑≤Ëß£Á∑®)`
          : highway.remark?.includes("Êú™Á¥çÁ∑®")
          ? `${highway.name} (Êú™Á¥çÁ∑®)`
          : highway.name}
      </Link>

      {showTooltip && hoveredHighway?.id === highway.id && (
        <div
          ref={cardRef}
          className={`absolute transition-opacity duration-150 ${
            visible ? "opacity-100" : "opacity-0"
          } ${
            position === "top" ? "bottom-full mb-4" : "top-full mt-4"
          } w-80 p-4 bg-white border border-gray-300 shadow-lg rounded-lg z-50 highway-card pointer-events-none`}
          style={{
            left: "50%",
            transform: "translateX(-50%)",
          }}
          onMouseEnter={handleCardMouseEnter}
          onMouseLeave={handleCardMouseLeave}
        >
          {/* ÂúñÁâáÂ±ïÁ§∫ */}
          {hoveredHighway.images?.length ? (
            <div className="relative">
              <img
                src={
                  hoveredHighway.images[hoveredHighway.currentImageIndex || 0]
                }
                alt={hoveredHighway.name}
                className="w-full h-48 object-cover rounded-md"
              />

              {hoveredHighway.images.length > 1 && (
                <>
                  <button
                    className={`absolute top-1/2 left-0 transform ${
                      hoveredHighway.currentImageIndex > 0 ? "block" : "hidden"
                    } -translate-y-1/2 bg-gray-800 text-white p-2 rounded-full z-10`}
                    onClick={handlePrevImage}
                    disabled={(hoveredHighway.currentImageIndex || 0) <= 0}
                  >
                    ‚óÄ
                  </button>

                  <button
                    className={`absolute top-1/2 right-0 transform ${
                      hoveredHighway.currentImageIndex <
                      hoveredHighway.images.length - 1
                        ? "block"
                        : "hidden"
                    } -translate-y-1/2 bg-gray-800 text-white p-2 rounded-full z-10`}
                    onClick={handleNextImage}
                    disabled={
                      (hoveredHighway.currentImageIndex || 0) >=
                      hoveredHighway.images.length - 1
                    }
                  >
                    ‚ñ∂
                  </button>
                </>
              )}

              <p className="text-sm text-black font-semibold mt-2">
                {(hoveredHighway.currentImageIndex || 0) + 1}/
                {hoveredHighway.images.length}
              </p>
            </div>
          ) : (
            <div className="w-full h-48 flex items-center justify-center bg-gray-200 rounded-md">
              <span className="text-gray-500">No image to show</span>
            </div>
          )}

          <p className="text-sm text-black font-semibold mt-2">
            {highway.name}
          </p>
          {highway.routeName && (
            <p className="text-sm text-black font-semibold mt-1">
              {highway.routeName}
            </p>
          )}
        </div>
      )}
    </h3>
  );
}
