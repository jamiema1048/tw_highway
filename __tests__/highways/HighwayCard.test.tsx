import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import HighwayCard from "../../src/app/highways/HighwayCard";
import { Highway } from "types/highway";
import { useState } from "react";

function Wrapper({ highway }: { highway: Highway }) {
  const [hoveredHighway, setHoveredHighway] = useState<Highway | null>(null);
  return (
    <HighwayCard
      highway={highway}
      hoveredHighway={hoveredHighway}
      setHoveredHighway={setHoveredHighway}
    />
  );
}

describe("HighwayCard", () => {
  const baseHighway: Highway = {
    id: 40100,
    name: "台1線",
    routeName: "台北－楓港",
    remark: "",
    images: ["https://placehold.co/400x200"],
  };

  it("hover link 時應該顯示卡片", async () => {
    render(<Wrapper highway={baseHighway} />);
    const link = screen.getByRole("link", { name: "台1線" });
    await userEvent.hover(link);
    await waitFor(() => {
      expect(screen.getByText("台北－楓港")).toBeInTheDocument();
    });
  });

  it("unhover link 時應該隱藏卡片", async () => {
    render(<Wrapper highway={baseHighway} />);
    const link = screen.getByRole("link", { name: "台1線" });
    await userEvent.hover(link);
    await userEvent.unhover(link);
    await waitFor(() => {
      expect(screen.queryByText("台北－楓港")).not.toBeInTheDocument();
    });
  });

  it("卡片位置應該不會超出螢幕邊界", async () => {
    Object.defineProperty(window, "innerWidth", { writable: true, value: 500 });
    render(<Wrapper highway={baseHighway} />);
    const link = screen.getByRole("link", { name: "台1線" });
    await userEvent.hover(link);

    const card = await screen.findByText("台北－楓港");
    await waitFor(() => {
      const style = window.getComputedStyle(card.parentElement!);
      expect(parseInt(style.left)).toBeGreaterThanOrEqual(0);
      expect(parseInt(style.left)).toBeLessThanOrEqual(500);
    });
  });

  it("長按 link (touchstart) 也會顯示卡片", async () => {
    vi.useFakeTimers();
    render(<Wrapper highway={baseHighway} />);
    const link = screen.getByRole("link", { name: "台1線" });

    fireEvent.touchStart(link);
    vi.advanceTimersByTime(600);

    await waitFor(() => {
      expect(screen.getByText("台北－楓港")).toBeInTheDocument();
    });

    fireEvent.touchEnd(link);
    vi.useRealTimers();
  });
});
