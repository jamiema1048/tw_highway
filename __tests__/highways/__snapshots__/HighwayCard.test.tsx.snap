// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HighwayCard > 應該正確渲染卡片 snapshot 1`] = `
<div
  class="p-4 bg-white border border-gray-300 shadow-lg rounded-lg highway-card"
  style="position: fixed; top: -9999px; left: -9999px; width: 20rem; z-index: 9999; transition: opacity 150ms; opacity: 0; pointer-events: none;"
>
  <div
    class="relative"
  >
    <img
      alt="台1線"
      class="w-full h-48 object-cover rounded-md"
      src="https://placehold.co/400x200"
    />
    <p
      class="text-sm text-black font-semibold mt-2"
    >
      1
      /
      1
    </p>
  </div>
  <p
    class="text-sm text-black font-semibold mt-2"
  >
    台1線
  </p>
  <p
    class="text-sm text-black font-semibold mt-1"
  >
    台北－楓港
  </p>
</div>
`;
